File: core/src/com/quillraven/game/core/ecs/system/ParticleSystem.java
package com.quillraven.game.core.ecs.system;

import com.badlogic.ashley.core.Entity;
import com.badlogic.ashley.core.Family;
import com.badlogic.ashley.systems.IteratingSystem;
import com.badlogic.gdx.graphics.g2d.ParticleEffect;
import com.badlogic.gdx.graphics.g2d.ParticleEffectPool;
import com.quillraven.game.core.Utils;
import com.quillraven.game.core.ecs.EntityEngine;
import com.quillraven.game.core.ecs.component.ParticleEffectComponent;

import java.util.EnumMap;

public class ParticleSystem extends IteratingSystem {
    private final EnumMap<ParticleEffectComponent.ParticleEffectType, ParticleEffectPool> effectPools;

    public ParticleSystem() {
        super(Family.all(ParticleEffectComponent.class).get());
        effectPools = new EnumMap<>(ParticleEffectComponent.ParticleEffectType.class);
    }

    @Override
    protected void processEntity(final Entity entity, final float deltaTime) {
        final ParticleEffectComponent peCmp = EntityEngine.peCmpMapper.get(entity);

        if (peCmp.effect != null) {
            // effect available -> update it
            peCmp.effect.update(deltaTime);
            if (peCmp.effect.isComplete()) {
                entity.remove(ParticleEffectComponent.class);
            }
        } else if (peCmp.type != ParticleEffectComponent.ParticleEffectType.NOT_DEFINED) {
            // type defined but effect not spawned yet -> spawn it
            ParticleEffectPool effectPool = effectPools.get(peCmp.type);
            if (effectPool == null) {
                final ParticleEffect effect = Utils.getResourceManager().get(peCmp.type.getEffectFilePath(), ParticleEffect.class);
                // set blend function cleanup to false to increase render performance because otherwise every effect
                // rendering will cause a spriteBatch flush.
                // this means that we need to take care of setting back the blend function by hand in the GameRenderSystem
                effect.setEmittersCleanUpBlendFunction(false);
                effectPool = new ParticleEffectPool(effect, 1, 128);
                effectPools.put(peCmp.type, effectPool);
            }
            peCmp.effect = effectPool.obtain();
            peCmp.effect.setPosition(peCmp.position.x, peCmp.position.y);
            peCmp.effect.scaleEffect(peCmp.scaling);
        }
    }
}

File: core/src/main/kotlin/com/github/quillraven/mysticgarden/system/ParticleSystem.kt
package com.github.quillraven.mysticgarden.system

import com.badlogic.gdx.graphics.OrthographicCamera
import com.badlogic.gdx.graphics.g2d.Batch
import com.badlogic.gdx.utils.viewport.Viewport
import com.github.quillraven.fleks.Entity
import com.github.quillraven.fleks.IteratingSystem
import com.github.quillraven.fleks.World.Companion.family
import com.github.quillraven.fleks.World.Companion.inject
import com.github.quillraven.mysticgarden.component.Boundary
import com.github.quillraven.mysticgarden.component.Particle
import ktx.graphics.use

class ParticleSystem(
    private val batch: Batch = inject(),
    private val gameViewport: Viewport = inject(),
    private val gameCamera: OrthographicCamera = inject(),
) : IteratingSystem(family { all(Particle, Boundary) }) {

    override fun onTick() {
        gameViewport.apply()
        batch.use(gameCamera) {
            val blndSrc = batch.blendSrcFunc
            val blndDst = batch.blendDstFunc

            super.onTick()

            // We need to manually restore blend functions because
            // we set 'setEmittersCleanUpBlendFunction' to false
            // for every effect to optimize draw calls
            batch.setBlendFunction(blndSrc, blndDst)
        }
    }

    override fun onTickEntity(entity: Entity) {
        val (x, y, w) = entity[Boundary]
        val (effect, offset, speed) = entity[Particle]

        with(effect) {
            setPosition(x + 0.5f * w + offset.x, y + offset.y)
            update(deltaTime * speed)
            draw(batch)
        }
    }
}